File: optimizers/NAdam.py
from .Optimizer import Optimizer, np


class NAdam(Optimizer):
    """
    NAdam — Adam с моментом Нестерова и затуханием момента.

    Первоисточник: Dozat T. "Incorporating Nesterov Momentum into Adam". ICLR 2016 Workshop.
    https://openreview.net/forum?id=OM0jvwB8jIp57ZJjtNEZ
    Реализация следует алгоритму из документации PyTorch:
    https://docs.pytorch.org/docs/stable/generated/torch.optim.NAdam.html
    """

    param_descriptions = {
        "lr": "скорость обучения — масштаб шага",
        "beta1": "коэффициент EMA первого момента (сглаживание направления градиента)",
        "beta2": "коэффициент EMA второго момента (усреднение квадратов градиента)",
        "eps": "малая добавка в знаменатель для численной устойчивости, настраивать обычно не нужно",
        "weight_decay": "L2-штраф: тянет точку к началу координат, искажая минимизируемую функцию",
        "momentum_decay": "темп прогрева нестеровского момента: чем больше, тем быстрее mu_t растёт к beta1",
    }

    v: np.ndarray
    acc: np.ndarray
    mu_product: float
    t: int

    def __init__(
        self,
        initial_x: np.ndarray,
        function,
        lr: float = 0.01,
        beta1: float = 0.9,
        beta2: float = 0.999,
        eps: float = 1e-08,
        weight_decay: float = 0.0,
        momentum_decay: float = 0.004,
    ) -> None:
        params = dict(
            lr=lr,
            beta1=beta1,
            beta2=beta2,
            eps=eps,
            weight_decay=weight_decay,
            momentum_decay=momentum_decay,
        )
        super().__init__(initial_x, function, params)

    def next_point(self) -> tuple[np.ndarray, float]:
        gradient = self.function.grad(self.x)
        self.t += 1
        gradient = gradient + self.params["weight_decay"] * self.x
        mu = self.params["beta1"] * (1 - 0.5 * 0.96 ** (self.t * self.params["momentum_decay"]))
        mu_next = self.params["beta1"] * (1 - 0.5 * 0.96 ** ((self.t + 1) * self.params["momentum_decay"]))
        self.mu_product *= mu
        mu_product_next = self.mu_product * mu_next
        self.v = self.params["beta1"] * self.v + (1 - self.params["beta1"]) * gradient
        self.acc = self.params["beta2"] * self.acc + (1 - self.params["beta2"]) * gradient**2
        bias_correction2 = 1 - self.params["beta2"] ** self.t
        denom = np.sqrt(self.acc / bias_correction2) + self.params["eps"]
        next_x = (
            self.x
            - self.params["lr"] * (1 - mu) / (1 - self.mu_product) * gradient / denom
            - self.params["lr"] * mu_next / (1 - mu_product_next) * self.v / denom
        )
        return self.move_next(next_x)

    def reset(self) -> None:
        super().reset()
        self.v = np.zeros([2])
        self.acc = np.zeros([2])
        self.mu_product = 1.0
        self.t = 0

File: src/shared/lib/optimization-engine/optimizers/nadam.ts
import type { OptimizerDescriptor } from "./types";
import { zero2, type Vec2 } from "@shared/lib/optimization-engine/linalg";
import { gradient } from "@shared/lib/optimization-engine/functions";

export const nAdamOptimizer: OptimizerDescriptor = {
  name: "NAdam",
  params: {
    lr: { default: 0.01, description: "скорость обучения — масштаб шага" },
    beta1: { default: 0.9, description: "коэффициент EMA первого момента (сглаживание направления градиента)" },
    beta2: { default: 0.999, description: "коэффициент EMA второго момента (усреднение квадратов градиента)" },
    eps: { default: 1e-8, description: "малая добавка в знаменатель для численной устойчивости, настраивать обычно не нужно" },
    weightDecay: {
      default: 0.0,
      description: "L2-штраф: тянет точку к началу координат, искажая минимизируемую функцию",
    },
    momentumDecay: {
      default: 0.004,
      description: "темп прогрева нестеровского момента: чем больше, тем быстрее mu_t растёт к beta1",
    },
  },
  createInstance(fn, initialX, params) {
    let x: Vec2 = initialX;
    let v: Vec2 = zero2();
    let acc: Vec2 = zero2();
    let muProduct = 1.0;
    let t = 0;
    return {
      get x() { return x; },
      params,
      next() {
        const rawGrad = gradient(fn, x[0], x[1]);
        t += 1;
        const grad: Vec2 = [rawGrad[0] + params.weightDecay * x[0], rawGrad[1] + params.weightDecay * x[1]];
        const mu = params.beta1 * (1 - 0.5 * 0.96 ** (t * params.momentumDecay));
        const muNext = params.beta1 * (1 - 0.5 * 0.96 ** ((t + 1) * params.momentumDecay));
        muProduct *= mu;
        const muProductNext = muProduct * muNext;
        v = [
          params.beta1 * v[0] + (1 - params.beta1) * grad[0],
          params.beta1 * v[1] + (1 - params.beta1) * grad[1],
        ];
        acc = [
          params.beta2 * acc[0] + (1 - params.beta2) * grad[0] ** 2,
          params.beta2 * acc[1] + (1 - params.beta2) * grad[1] ** 2,
        ];
        const biasCorrection2 = 1 - params.beta2 ** t;
        const denom: Vec2 = [
          Math.sqrt(acc[0] / biasCorrection2) + params.eps,
          Math.sqrt(acc[1] / biasCorrection2) + params.eps,
        ];
        x = [
          x[0]
            - (params.lr * (1 - mu)) / (1 - muProduct) * (grad[0] / denom[0])
            - (params.lr * muNext) / (1 - muProductNext) * (v[0] / denom[0]),
          x[1]
            - (params.lr * (1 - mu)) / (1 - muProduct) * (grad[1] / denom[1])
            - (params.lr * muNext) / (1 - muProductNext) * (v[1] / denom[1]),
        ];
        return { x, value: fn(x[0], x[1]) };
      },
      reset() {
        x = initialX;
        v = zero2();
        acc = zero2();
        muProduct = 1.0;
        t = 0;
      },
    };
  },
};
